alf/algorithms/algorithm.py

```diff
@@ -188,21 +187,54 @@ def default_optimizer(self):
         """Get the default optimizer for this algorithm."""
         return self._default_optimizer
 
+    def _assert_no_cycle(self):
+        visited = set()
+        to_be_visited = [self]
+        while to_be_visited:
+            node = to_be_visited.pop(0)
+            visited.add(node)
+            for child in node._get_children():
+                assert child not in visited, (
+                    "There is a cycle in the "
+                    "algorithm tree caused by '%s'" % child.name)
+                if isinstance(child, Algorithm):
+                    to_be_visited.append(child)
```

Comment:
I don't think this BFS will work, because even when a child at two levels lower is a duplicate of a node at the current level, by the time we visit that child, all the current-level nodes will have been removed from the queue.